extensions/amp-user-notification/0.1/amp-user-notification.js

```diff
@@ -0,0 +1,423 @@
+   * @param {!PostResponseMetadata}
+   * @return {!Promise<boolean>}
+   * @private
+   */
+  onPostShowEndpointSuccess_(data) {
+    assert(typeof data['showNotification'] == 'boolean', '`showNotification` ' +
+        'should be a boolean. Got "%s" which is of type %s.',
+        data['showNotification'], typeof data['showNotification']);
+
+    if (data['showNotification']) {
+      return data['showNotification'];
+    }
+    // if no notification is shown, resolve right away with `false` since
+    // we dont need to show it.
+    return this.dialogResolve_(false);
+  }
+
+  /**
+   * Generic rejection handler for amp-user-notification.
+   * @return {boolean}
+   * @private
+   */
+  onRejection_(msg) {
+    log.error(TAG_, 'An error ocurred on `amp-user-notification` ' +
+        `with html element id "${this.elementId_}". message: ${msg}`);
+    return false;
+  }
+
+  /**
+   * Get async cid service.
+   * @return {!Promise}
+   * @private
+   */
+  getAsyncCid_() {
+    return cidFor(this.win_).then(cid => {
+      // `amp-user-notification` is our cid scope, while we give it a resolved
+      // promise for the 2nd argument so that the 3rd argument (the
+      // persistentConsent) is the one used to resolve getting
+      // the external CID.
+      return cid.get('amp-user-notification',
+          Promise.resolve(), this.dialogPromise_);
+    });
+  }
+
+  /**
+   * @return {!Promise<boolean>}
+   * @private
+   **/
+  getIfHref_() {
+    if (!this.getIfHrefPromise_) {
+      this.getIfHrefPromise_ = this.getAsyncCid_()
+          .then(this.postShowEndpoint_.bind(this))
+          .then(this.onPostShowEndpointSuccess_.bind(this))
+          .catch(this.onRejection_.bind(this));
+    }
+    return this.getIfHrefPromise_;
+  }
+
+  /** @override */
+  shouldShow() {
+    return this.getIfHref_();
+  }
+
+  /**
+   * Shows the current user notification. Adds the `.amp-active` class to
+   * the element.
+   */
+  show() {
+    this.element.style.display = 'block';
+    this.element.classList.add('amp-active');
+  }
+
+  /**
+   * Hides the current user notification and invokes the `dialogResolve_`
+   * method. Removes the `.amp-active` class from the element.
+   */
+  dismiss() {
+    this.element.style.display = 'none';
+    this.element.classList.remove('amp-active');
+    this.element.classList.add('amp-hidden');
+    this.dialogResolve_(false);
+    this.postDismissEnpoint_();
+  }
+
+  /** @override */
+  isDismissed() {
+    return this.dialogPromise_;
+  }
+}
+
+
+/**
+ * UserNotificationService handles `amp-user-notification`
+ * queuing and registration, as well as exposing the components
+ * dismiss promise.
+ */
+export class UserNotificationService {
+
+  constructor(window) {
+    this.win_ = window;
+
+    /** @private @const {!Object<!UserNotificationDefer>} */
+    this.deferRegistry_ = Object.create(null);
+
+    /** @private @const {!Viewer} */
+    this.viewer_ = viewerFor(this.win_);
```

Comment:
Probably a good idea, but the size concern goes away when we land the service-implementation-split changes.